tensorflow/core/kernels/mkl_qmatmul_op.cc

```diff
@@ -0,0 +1,870 @@
+
+      std::vector<mkldnn::primitive> net;
+      float out_scale;
+      // If the bias is float and input quantize is MIN_FIRST, bias has to be
+      // compensated with B's32 = Q'a * Qw * B𝑓32 + Q'a * Qw * Min(A𝑓32) * 1 *
+      // W𝑓32.
+      if (mode_ == QUANTIZE_MODE_MIN_FIRST) {
+        int k = weight_tensor.dim_size(0);
+        int n = weight_tensor.dim_size(1);
+        float* comp_bias = GetCompBiasBuffer(n);
+
+        qint8* wt_buf = static_cast<qint8*>(
+            const_cast<qint8*>(weight_tensor.flat<qint8>().data()));
+
+        const float* bias_buf = static_cast<float*>(
+            const_cast<float*>(bias_tensor.flat<float>().data()));
+
+        float qa_amin = 255 * min_input / (max_input - min_input);
+
+        out_scale = (255.0 * 127.0) /
+                    ((max_input - min_input) *
+                     std::max(std::abs(max_weight), std::abs(min_weight)));
+
+#pragma omp parallel for schedule(static)
+        for (int j = 0; j < n; ++j) {
+          int x = 0;
+          for (int i = 0; i < k; ++i) {
+            x += wt_buf[i * n + j];
+          }
+          comp_bias[j] =
+              ((bias_buf[j] * out_scale) + static_cast<float>(x * qa_amin));
+        }
+
+        return reinterpret_cast<Tbias*>(comp_bias_);
+
+      } else if (mode_ == QUANTIZE_MODE_SCALED) {
+        // If the bias is float and input quantize is SCALE, bias has to be
+        // compensated with Bs32 = Qa * Qw * Bf32.
+        out_scale = 255.0 * 127.0 / max_input *
+                    std::max(std::abs(max_weight), std::abs(min_weight));
+
+        std::vector<float> scales;
+        scales.push_back(out_scale);
+        mkldnn::primitive_attr bias_attr;
+        bias_attr.set_output_scales(0, scales);
+
+        void* bias_buf = static_cast<void*>(
+            const_cast<Tbias*>(bias_tensor.flat<Tbias>().data()));
+        input_bias_ =
+            new memory(mkldnn_matmul_fwd_pd->bias_primitive_desc(), bias_buf);
+        scaled_bias_ = new memory(mkldnn_matmul_fwd_pd->bias_primitive_desc());
+        auto reorder_desc = mkldnn::reorder::primitive_desc(
+            input_bias_->get_primitive_desc(),
+            scaled_bias_->get_primitive_desc(), bias_attr);
+        net.push_back(
+            mkldnn::reorder(reorder_desc, *input_bias_, *scaled_bias_));
+        stream(stream::kind::eager).submit(net).wait();
+        return reinterpret_cast<Tbias*>(scaled_bias_->get_data_handle());
+      } else {
+        context->CtxFailure(
+            errors::InvalidArgument("Quantization mode must be"
+                                    "either MIN_FIRST or SCALED."));
+      }
+    }
+  }
+
+  // Allocate output tensor.
+  virtual void AllocateOutputTensor(
+      OpKernelContext* context,
+      const inner_product_forward::primitive_desc& mkldnn_matmul_prim_desc,
+      const memory::dims& output_dims_mkl_order,
+      memory::format output_tf_format, Tensor** output_tensor) {
+    CHECK_NOTNULL(output_tensor);
```

Comment:
Sorry I missed this earlier. We don't allow `CHECK` in new code anymore. Please use `DCHECK` instead.
```suggestion
    DCHECK(output_tensor);
```

Reply:
@penpornk changed to DCHECK.